drivers/net/adin1110/adin1110.c

```diff
@@ -0,0 +1,820 @@
+
+	ret = adin1110_reg_write(desc, ADIN1110_MDIOACC(0), val);
+	if (ret)
+		return ret;
+
+	/* The PHY will set the MDIO_TRDONE bit to 1 once a transaction completes */
+	while (!mdio_done) {
+		ret = adin1110_reg_read(desc, ADIN1110_MDIOACC(0), &mdio_val);
+		if (ret)
+			return ret;
+
+		mdio_done = no_os_field_get(ADIN1110_MDIO_TRDONE, mdio_val);
+	}
+
+	*data = no_os_field_get(ADIN1110_MDIO_DATA, mdio_val);
+
+	return 0;
+}
+
+/**
+ * @brief Write a PHY register using clause 22
+ * @param desc - the device descriptor
+ * @param phy_id - the phy device's id
+ * @param reg - register's address
+ * @param data - register's value
+ * @return 0 in case of success, negative error code otherwise
+ */
+int adin1110_mdio_write(struct adin1110_desc *desc, uint32_t phy_id,
+			uint32_t reg, uint32_t data)
+{
+	uint32_t mdio_val = 0;
+	uint32_t mdio_done = 0;
+	uint32_t val;
+	int ret;
+
+	/* Use clause 22 for the MDIO register addressing */
+	val = no_os_field_prep(ADIN1110_MDIO_ST, 0x1);
+	val |= no_os_field_prep(ADIN1110_MDIO_OP, ADIN1110_MDIO_OP_WR);
+	val |= no_os_field_prep(ADIN1110_MDIO_PRTAD, phy_id);
+	val |= no_os_field_prep(ADIN1110_MDIO_DEVAD, reg);
+	val |= no_os_field_prep(ADIN1110_MDIO_DATA, data);
+
+	ret = adin1110_reg_write(desc, ADIN1110_MDIOACC(0), val);
+	if (ret)
+		return ret;
+
+	/* The PHY will set the MDIO_TRDONE bit to 1 once a transaction completes */
+	while (!mdio_done) {
+		ret = adin1110_reg_read(desc, ADIN1110_MDIOACC(0), &mdio_val);
+		if (ret)
+			return ret;
+
+		mdio_done = no_os_field_get(ADIN1110_MDIO_TRDONE, mdio_val);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Write a PHY register using clause 45
+ * @param desc - the device descriptor
+ * @param phy_id - the phy device's MDIO id
+ * @param dev_id - the device id of the register
+ * @param reg - register's address
+ * @param data - register's value
+ * @return 0 in case of success, negative error code otherwise
+ */
+int adin1110_mdio_write_c45(struct adin1110_desc *desc, uint32_t phy_id,
+			    uint32_t dev_id, uint32_t reg, uint32_t data)
+{
+	uint32_t mdio_val = 0;
+	uint32_t mdio_done = 0;
+	uint32_t val;
+	int ret;
+
+	val = no_os_field_prep(ADIN1110_MDIO_ST, 0x0);
+	val |= no_os_field_prep(ADIN1110_MDIO_OP, ADIN1110_MDIO_OP_ADDR);
+	val |= no_os_field_prep(ADIN1110_MDIO_PRTAD, phy_id);
+	val |= no_os_field_prep(ADIN1110_MDIO_DEVAD, dev_id);
+	val |= no_os_field_prep(ADIN1110_MDIO_DATA, reg);
+
+	ret = adin1110_reg_write(desc, ADIN1110_MDIOACC(0), val);
+	if (ret)
+		return ret;
+
+	val = no_os_field_prep(ADIN1110_MDIO_ST, 0x0);
+	val |= no_os_field_prep(ADIN1110_MDIO_OP, ADIN1110_MDIO_OP_WR);
+	val |= no_os_field_prep(ADIN1110_MDIO_PRTAD, phy_id);
+	val |= no_os_field_prep(ADIN1110_MDIO_DEVAD, dev_id);
+	val |= no_os_field_prep(ADIN1110_MDIO_DATA, data);
+
+	ret = adin1110_reg_write(desc, ADIN1110_MDIOACC(1), val);
+	if (ret)
+		return ret;
+
+	/* The PHY will set the MDIO_TRDONE bit to 1 once a transaction completes */
+	while (!mdio_done) {
```

Comment:
this while loop seems to be occurring multiple times. I'd add a separate parametrized function for this.